### ISSUE TITLE
brick/math method for gmp_invert

### ISSUE THREAD

**Issue by Spomky** [brick/math#47]
Hi,

During the implementation of this library in one of my projects, I spent a few moment to find a way to use the GMP method [`gmp_invert`](https://www.php.net/manual/fr/function.gmp-invert.php).
In the end, I created my own method, by it could be directly implemented here: the gmp engine will directly use the gmp_invert method. Other engines can use the following.

Hope it helps.

```php
<?php

declare(strict_types=1);

require_once 'vendor/autoload.php';

use Brick\Math\BigInteger;
use Brick\Math\RoundingMode;

function modInverse(BigInteger $a, BigInteger $b): BigInteger
{
    $x = BigInteger::zero();
    $y = BigInteger::zero();
    $g = gcdExtended($a, $b, $x, $y);
    if (!$g->isEqualTo(BigInteger::one())) {
        throw new InvalidArgumentException('Unable to compute the modInverse for the given modulus');
    }

    return $x->mod($b)->plus($b)->mod($b);
}

function gcdExtended(BigInteger $a, BigInteger $b, BigInteger &$x, BigInteger &$y): BigInteger
{
    if ($a->isEqualTo(BigInteger::zero())) {
        $x = BigInteger::zero();
        $y = BigInteger::one();

        return $b;
    }

    $x1 = BigInteger::zero();
    $y1 = BigInteger::zero();
    $gcd = gcdExtended($b->mod($a), $a, $x1, $y1);

    $x = $y1->minus($b->dividedBy($a, RoundingMode::FLOOR)->multipliedBy($x1));
    $y = $x1;

    return $gcd;
}
```

Example:

```php
gmp_invert(5, 11); //GMP resource of value 9

$five = BigInteger::of(5);
$eleven = BigInteger::of(11);
modInverse($five, $eleven); // BigInteger of value 9
```

**Comment by BenMorel** [comment]
Hi, thank you for the feature request and for providing working code! 👍 

I'll take some time in the coming days to implement it in the library and add some tests.

Question: is there any value in providing `gcdExtended()` as a public method? Or is it just an implementation detail, with no real value on its own? I can see that [Java's BigInteger](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/math/BigInteger.html) has a `modInverse()` method, but no `gcdExtended()`.

Finally, out of curiosity, I'd be interested to know what's your use case for `modInverse()`!

**Comment by Spomky** [comment]
Hi,

I am happy to read that!
The method `gcdExtended` can be private. At the moment, I don't see any reason for using it directly.
In my context, this function is used for the computation of the [Chinese Remainder Theorem (CRT)](https://en.wikipedia.org/wiki/Chinese_remainder_theorem) and especially for [RSA encryption algorithm](https://en.wikipedia.org/wiki/RSA_(cryptosystem)) (the CRT reduces the computation time).

**Comment by BenMorel** [comment]
Hi, I just started implementing `modInverse()`, however it looks like the algorithm above fails for negative numbers:

```php
gmp_invert('1234567890', '-19'); // 11
modInverse('1234567890', '-19'); // -8

gmp_invert('-1234567890', '19'); // 8
modInverse('-1234567890', '19'); // exception

gmp_invert('-1234567890', '-19'); // 8
modInverse('-1234567890', '-19'); // exception
```

Any idea how to fix it?

**Comment by BenMorel** [comment]
> Any idea how to fix it?

No worries, I got the right result by borrowing [this solution](https://github.com/openjdk/jdk/blob/99c7b2b85b0d3bd138d4dcad083a6ecf1de30b3d/src/java.base/share/classes/java/math/BigInteger.java#L3281-L3282) from the JDK.

**Comment by BenMorel** [comment]
Now the question is, what should we do when `$m` is negative? Java & Dart throw an exception, while `gmp_invert()` returns the value of modInverse($x, abs($m)). For now I've allowed negative modulo this way, but I'm not sure this is the right way the API should respond.

**Comment by BenMorel** [comment]
Otherwise looks good to me now. Could you please test the `master` branch (with and without GMP please!) and report any issue?

**Comment by BenMorel** [comment]
With no feedback received, I've moved forward with throwing a `NegativeNumberException` when the modulus is negative.

Now available as `modInverse()` and released as [0.8.16](https://github.com/brick/math/releases/tag/0.8.16)! ✨
